[Liquid Glass] indiatimes.com: opening the hamburger menu before scrolling causes the top scroll pocket to disappear

whsieh opened the pull request:
#### d146201b066c48c4254e948676a683a02bfa3402
<pre>
[Liquid Glass] indiatimes.com: opening the hamburger menu before scrolling causes the top scroll pocket to disappear
<a href="https://bugs.webkit.org/show_bug.cgi?id=297020">https://bugs.webkit.org/show_bug.cgi?id=297020</a>
<a href="https://rdar.apple.com/157665752">rdar://157665752</a>

Reviewed by Abrar Rahman Protyasha and Megan Gardner.

When opening the hamburger menu on indiatimes.com without having scrolled the page, a dimming layer
appears over the page, which the heuristic in `LocalFrameView::fixedContainerEdges` correctly
identifies. We then attempt to take the sampled color (black at ~50% opacity) and use it as the
fixed color extension and solid color hard pocket background.

The problem with this is that `UIScrollView` internally rejects any rejects any explicitly-set
capture color that has an alpha &lt; 1, instead treating the color as `nil`. This in turn breaks luma
tracking in the scroll pocket, resulting in an unexpected solid white top pocket in dark mode. To
mitigate this, we ensure that even in the modal &quot;dimming layer&quot; case, we always return an opaque
fixed color by blending the dimming layer&apos;s background color against the page background color (so
in this case, `rgba(0, 0, 0, 0.5)` against the white page background yields solid medium-gray).

It&apos;s also somewhat unexpected that the top scroll pocket background color changes as a result of the
user opening a menu — while our current policy prevents the pocket from changing after the user has
scrolled the page, it should probably extend to after the user has started interacting with the page
by clicking or tapping.

We fix both of these issues in this patch.

* LayoutTests/fast/page-color-sampling/blend-modal-background-color-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/blend-modal-background-color.html: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):

See above.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUserDidInteractWithPage):
(WebCore::Page::didCommitLoad):
(WebCore::Page::updateFixedContainerEdges):

In addition to user-triggered scrolling, make it so that any other interaction with the page (i.e.
anything that triggers a user gesture token) will also prevent the sampled color from changing. This
prevents us from picking up the color change at all, in this case.

* Source/WebCore/page/Page.h:

Add a new flag to keep track of whether the user has interacted with the page, since the last load
committed.

Canonical link: <a href="https://commits.webkit.org/298351@main">https://commits.webkit.org/298351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4add3701b18af3de0d1cf1a6a3b71ffc1811b035

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34927 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25419 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121332 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65838 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f4eee91d-e6c3-498b-af38-f846d4e958a0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35578 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43493 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87553 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f7329925-a847-4147-bc2b-67281d92f87f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28366 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103439 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67950 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27538 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64982 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97751 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21676 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124511 "Built successfully") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42179 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31562 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96348 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42550 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99629 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96135 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24460 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41346 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19192 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38138 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42057 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47604 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41583 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44907 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43311 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7571efa7390a2bd2a6bbc96ed2b15eb221d25eff. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49031

aprotyas commented on LayoutTests/fast/page-color-sampling/blend-modal-background-color.html:
```diff
@@ -0,0 +1,45 @@
+    jsTestIsAsync = true;
+
+    addEventListener("load", async () => {
+        await UIHelper.ensurePresentationUpdate();
```
Do you need to wait for an animation frame here? Just thinking about the test failure

whsieh replied:
Oh, adding `await UIHelper.setObscuredInsets(100, 0, 0, 0);` seems to have fixed the macOS test failure (strangely, it was only failing after the first iteration of that test for me locally).

I think there's a bug in the test harness regarding the `obscuredInset.top` test option — I should probably investigate that in a followup…

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1083f7c5aa009e3dd143615336bec9f4fcc38b5f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49031

whsieh commented:
Thanks for the reviews!

whsieh committed:
[Liquid Glass] indiatimes.com: opening the hamburger menu before scrolling causes the top scroll pocket to disappear
https://bugs.webkit.org/show_bug.cgi?id=297020
rdar://157665752

Reviewed by Abrar Rahman Protyasha and Megan Gardner.

When opening the hamburger menu on indiatimes.com without having scrolled the page, a dimming layer
appears over the page, which the heuristic in `LocalFrameView::fixedContainerEdges` correctly
identifies. We then attempt to take the sampled color (black at ~50% opacity) and use it as the
fixed color extension and solid color hard pocket background.

The problem with this is that `UIScrollView` internally rejects any rejects any explicitly-set
capture color that has an alpha < 1, instead treating the color as `nil`. This in turn breaks luma
tracking in the scroll pocket, resulting in an unexpected solid white top pocket in dark mode. To
mitigate this, we ensure that even in the modal "dimming layer" case, we always return an opaque
fixed color by blending the dimming layer's background color against the page background color (so
in this case, `rgba(0, 0, 0, 0.5)` against the white page background yields solid medium-gray).

It's also somewhat unexpected that the top scroll pocket background color changes as a result of the
user opening a menu — while our current policy prevents the pocket from changing after the user has
scrolled the page, it should probably extend to after the user has started interacting with the page
by clicking or tapping.

We fix both of these issues in this patch.

* LayoutTests/fast/page-color-sampling/blend-modal-background-color-expected.txt: Added.
* LayoutTests/fast/page-color-sampling/blend-modal-background-color.html: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):

See above.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUserDidInteractWithPage):
(WebCore::Page::didCommitLoad):
(WebCore::Page::updateFixedContainerEdges):

In addition to user-triggered scrolling, make it so that any other interaction with the page (i.e.
anything that triggers a user gesture token) will also prevent the sampled color from changing. This
prevents us from picking up the color change at all, in this case.

* Source/WebCore/page/Page.h:

Add a new flag to keep track of whether the user has interacted with the page, since the last load
committed.

Canonical link: https://commits.webkit.org/298351@main
